packages/devtools/src/devtools.ts

```diff
@@ -0,0 +1,448 @@
+
+type QueryEventTypes = {
+  'queries:start': {
+    query: string;
+    materialized: boolean;
+    clientId?: string;
+    timestamp: number;
+  };
+  'queries:end': {
+    query: string;
+    materialized: boolean;
+    clientId?: string;
+    latencyMs?: number;
+    numRows?: number;
+    numBytes?: number;
+    exemplarRows?: Record<string, unknown>[]; // first ~10 rows
+    timestamp: number;
+  };
+};
+
+type ClientState = {
+  clientId: string;
+  connected: boolean;
+  lastQuery?: string;
+  lastLatencyMs?: number;
+  lastNumRows?: number;
+  lastNumBytes?: number;
+  lastExemplarRows?: Record<string, unknown>[];
+  children?: ClientState[]; // for tree structure
+};
+
+type ClientEventTypes = {
+  'clients:state': {
+    clients: ClientState[];
+    coordinatorState: Record<string, unknown>;
+    numConnectedClients: number;
+    timestamp: number;
+  };
+};
+
+// Module-level singletons for the event clients
+let _queryEventClient: EventClient<QueryEventTypes> | null = null;
+let _clientEventClient: EventClient<ClientEventTypes> | null = null;
+
+/**
+ * Return (and lazily create) the TanStack event clients.
+ */
+function createEventClients() {
+  if (!_queryEventClient || !_clientEventClient) {
+    _queryEventClient = new EventClient<QueryEventTypes>({
+      pluginId: 'queries',
+      debug: true
+    });
+    _clientEventClient = new EventClient<ClientEventTypes>({
+      pluginId: 'clients',
+      debug: true
+    });
+  }
+  return {
+    queryEventClient: _queryEventClient,
+    clientEventClient: _clientEventClient
+  };
+}
+
+/**
+ * Bridge function that connects core AsyncDispatch events to TanStack EventClient.
+ * This subscribes to core events and forwards them to TanStack devtools.
+ */
+function bridgeEvents() {
+  const { queryEventClient, clientEventClient } = createEventClients();
+
+  // Bridge query start events
+  queryEvents.addEventListener('start', (payload) => {
+    queryEventClient.emit('start', payload);
+  });
+
+  // Bridge query end events
+  queryEvents.addEventListener('end', (payload) => {
+    queryEventClient.emit('end', payload);
+  });
+
+  // Bridge client state events
+  clientEvents.addEventListener('state', (payload) => {
+    clientEventClient.emit('state', payload);
+  });
+}
+
+/**
+ * Initialize a very small devtools UI and hook it up to the shared EventClients.
+ * The plugin will append nodes into the element you mount it into and update
+ * when events arrive.
+ */
+export function initializeDevtools(document: Document) {
+  // Bridge core events to TanStack EventClient
+  bridgeEvents();
+
+  // avoid adding multiple devtools containers
+  let devtoolsDiv = document.getElementById('devtools');
+  if (!devtoolsDiv) {
+    devtoolsDiv = document.createElement('div');
+    devtoolsDiv.id = 'devtools';
+    document.body.appendChild(devtoolsDiv);
+  }
+
+  const { queryEventClient, clientEventClient } = createEventClients();
+
+  const devtools = new TanStackDevtoolsCore({
```

Comment:
Most of this seems to be about logging which browser consoles already support pretty well (with console.log and even grouping etc (e.g. https://developer.mozilla.org/en-US/docs/Web/API/console/group_static).